builtin/logical/pki/path_tidy.go

```diff
@@ -905,122 +938,136 @@ func (b *backend) startTidyOperation(req *logical.Request, config *tidyConfig) {
 }
 
 func (b *backend) doTidyCertStore(ctx context.Context, req *logical.Request, logger hclog.Logger, config *tidyConfig) (uint, error) {
-	serials, err := req.Storage.List(ctx, "certs/")
-	if err != nil {
-		return 0, fmt.Errorf("error fetching list of certs: %w", err)
-	}
-
-	revokedSafetyBuffer := config.SafetyBuffer
-	if config.RevokedSafetyBuffer != nil {
-		revokedSafetyBuffer = *config.RevokedSafetyBuffer
-	}
-
-	haveWarned := false
-	serialCount := len(serials)
-	metrics.SetGauge([]string{"secrets", "pki", "tidy", "cert_store_total_entries"}, float32(serialCount))
-
+	var after string
+	var serialCount uint
 	var revokedDeleted uint
-	for i, serial := range serials {
-		b.tidyStatusMessage(fmt.Sprintf("Tidying certificate store: checking entry %d of %d", i, serialCount))
-		metrics.SetGauge([]string{"secrets", "pki", "tidy", "cert_store_current_entry"}, float32(i))
+	haveWarned := false
 
-		// Check for cancel before continuing.
-		if atomic.CompareAndSwapUint32(b.tidyCancelCAS, 1, 0) {
-			return 0, tidyCancelledError
+	for {
+		// Fetch the next batch of certificate serials using pagination
+		serials, err := req.Storage.ListPage(ctx, "certs/", after, config.PageSize)
+		if err != nil {
+			return 0, fmt.Errorf("error fetching list of certs: %w", err)
 		}
 
-		// Check for pause duration to reduce resource consumption.
-		if config.PauseDuration > (0 * time.Second) {
-			time.Sleep(config.PauseDuration)
+		// If no serials are returned, we've reached the end of the cert list
+		if len(serials) == 0 {
+			break
 		}
 
-		certEntry, err := req.Storage.Get(ctx, "certs/"+serial)
-		if err != nil {
-			return 0, fmt.Errorf("error fetching certificate %q: %w", serial, err)
+		serialCount += uint(len(serials))
+
+		revokedSafetyBuffer := config.SafetyBuffer
+		if config.RevokedSafetyBuffer != nil {
+			revokedSafetyBuffer = *config.RevokedSafetyBuffer
 		}
 
-		if certEntry == nil {
-			logger.Warn("certificate entry is nil; tidying up since it is no longer useful for any server operations", "serial", serial)
-			if err := req.Storage.Delete(ctx, "certs/"+serial); err != nil {
-				return 0, fmt.Errorf("error deleting nil entry with serial %s: %w", serial, err)
+		for i, serial := range serials {
+			// Set the `after` variable for the next ListPage call
+			after = serial
+
+			b.tidyStatusMessage(fmt.Sprintf("Tidying certificate store: checking entry %d of %d on current page; total certs checked: %d", i, config.PageSize, int(serialCount)+i))
+			metrics.SetGauge([]string{"secrets", "pki", "tidy", "cert_store_current_entry"}, float32(int(serialCount)+i-1))
```

Comment:
I'm also confused why you have a `-1` on this...? I think it should just be the same value as in the status line above.

Reply:
`TestBackend_RevokePlusTidy_Intermediate` was failing and after I looked at it and the value it was expecting for `cert_store_current_entry`, I thought it was erring because counting starts at 0 not 1 (which is true but this is not the correct solution if they are many entries), hence the `-1`. Moving the `serialCount` to below the for loop fixed this problem.

Reply:
@fatima2003 Ahh right, because it isn't a count, it is the index. My bad!

Reply:
The way it was written was a tad bit confusing ><